src/midi/midi_fluidsynth.cpp

```diff
@@ -375,17 +394,47 @@ void MidiHandlerFluidsynth::PrintStats()
 
 void MidiHandlerFluidsynth::MixerCallBack(uint16_t requested_frames)
 {
-	constexpr uint16_t max_frames = expected_max_frames; // local fixes link error
-	std::vector<float> render_buffer(max_frames * 2);    // L & R channels
+	while (requested_frames) {
+		const auto frames_to_be_played = std::min(GetRemainingFrames(),
+		                                          requested_frames);
+		const auto sample_offset_in_buffer = play_buffer.data() +
+		                                     last_played_frame * 2;
 
-	while (requested_frames > 0) {
-		const uint16_t frames = std::min(requested_frames, max_frames);
+		assert(frames_to_be_played <= play_buffer.size());
+		channel->AddSamples_s16(frames_to_be_played, sample_offset_in_buffer);
 
-		fluid_synth_write_float(synth.get(), frames, render_buffer.data(),
-		                        0, 2, render_buffer.data(), 1, 2);
-		const auto out = soft_limiter.Process(render_buffer, frames);
-		channel->AddSamples_s16(frames, out.data());
-		requested_frames -= frames;
+		requested_frames -= frames_to_be_played;
+		last_played_frame += frames_to_be_played;
+	}
+}
+
+// GetRemainingFrames() gives us the number of unplayed (ie: remaining) frames
+// left in the current play buffer. If it's run out, then we pop a new "full"
+// play buffer out of the ring. If the ring doesn't have any  to offer, then we
+// wait until it does.
+uint16_t MidiHandlerFluidsynth::GetRemainingFrames()
+{
+	// the current play buffer has some frames left, so return those
+	if (last_played_frame < FRAMES_PER_BUFFER)
+		return FRAMES_PER_BUFFER - last_played_frame;
+
+	// Otherwise move the next buffer out of the ring into our play buffer
+	ring.wait_dequeue(play_buffer);
+	last_played_frame = 0;
+
+	return FRAMES_PER_BUFFER;
+}
+
+// After GetRemainingFrames() pops out a buffer for playing, back-pressure is
+// released in the ring allowing FluidSynth to renderer the next buffer.
+void MidiHandlerFluidsynth::Render()
+{
+	std::vector<float> render_buffer(FRAMES_PER_BUFFER * 2); // L & R channels
+	while (keep_rendering) {
+		fluid_synth_write_float(synth.get(), FRAMES_PER_BUFFER,
+		                        render_buffer.data(), 0, 2, render_buffer.data(), 1, 2);
+		auto out = soft_limiter.Process(render_buffer, FRAMES_PER_BUFFER);
+		ring.wait_enqueue(out);
```

Comment:
The more I think about this solution, the more problematic it becomes…

So we have a rwqueue that can handle up to 8 "buffers". No more. And we keep churning-in - allocating, filling with data, and pushing new vectors in there again and again… moving vectors is not a problem - that happens quickly; we might be wasting resources with re-allocating and freeing vectors over and over again (when it's totally unnecessary).

*Do not change this yet. Let's merge this in - after merge we'll have at least the same solution in both MT-32 and FluidSynth. After that we'll think what's the best move forward and we both will be able to do some testing.*

Reply:
You're description was brilliant here @dreamer " we keep churning-in - allocating, filling with data, and pushing new vectors in there again and again… moving vectors is not a problem - that happens quickly; we might be wasting resources with re-allocating and freeing vectors over and over again (when it's totally unnecessary)."

Fully agree - I think there's one last optimization step that will nail this:

1) We know the rendering phase into the vector<float>'s already good, because it only uses a single vector for the entire lifetime.
2) The re-allocation happens in the SoftLimiter, where it creates the vector<int16_t> temporary, which is eventually moved into the queue, moved out of the queue, and then eventually destroyed.

If we passed a `& vector<int16_t>` into the SoftLimiter for it to write into, then we could keep recycling it on the outside.  

> Do not change this yet. Let's merge this in - after merge we'll have at least the same solution in both MT-32 and FluidSynth. 

Absolutely :-)  We've got a working baseline, and you're headed to take it to the next level!